pkg/kubelet/cm/dra/state/checkpoint/v1alpha1/types.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"encoding/json"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager/checksum"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type DRACheckpoint struct {
+	metav1.TypeMeta
+	Data     []ClaimInfoState
+	Checksum checksum.Checksum
+}
```

Comment:
Is the idea that future versions will have the same field names here, but the implementation of `Data` might be different? And we know how to decode `Data` based on the API group / kind from `meta`?

Reply:
Yes, that's the idea. DRACheckpoint structure could also be different as far as I understand. @pohly is it correct?

Reply:
Yes.

Where I am not following (but I haven't looked at your latest code) is how and when the `Checksum` is going to be calculated.

Is it over the in-memory `Data` content? Then a reader has to support exactly the version that the writer was using, which somewhat defeats the purpose of having a self-describing file content with conversion. For example, are downgrades meant to be supported? I think they have to be.

That leads to the following problem:
- kubelet v1.33 writes a file using format V2.
- Node gets downgraded to v1.32, which doesn't support V2.
- kubelet v1.32 cannot read the file (no registered type for V2, no way to verify the checksum).

Here's how I would do it:
- Define a "v1" format. Always use that version for the file content.
- When extending the data, add new fields to that version. Those fields must be something that older kubelets can ignore.
- Don't bother with conversion. We only need it if, for whatever reason, we want to change the type of the data such that adding new fields isn't sufficient. Even then we would still need to extend v1 so that we can continue to use it as storage version.
- Calculate checksums after encoding as JSON, check it before decoding. This can be done by having a simpler container struct:
```
type Checkpoint struct {
    Data string // JSON representation of v1.DRACheckpoint
    Checksum checksum.Checksum // hash of Data string
}
```

Reply:
Note that I am skipping adding a version and kind to that container struct. We basically have define it once and then can never change it. The only change that *might* occur is changing the checksumming, but that then would break downgrades again and would need to be handled carefully.

Reply:
@klueska checksum is calculated over jsoned checkpoint. Look at `pkg/kubelet/cm/dra/checkpoint/v1/apis.go:MarshalCheckpoint` for more details.

Reply:
> When extending the data, add new fields to that version. Those fields must be something that older kubelets can ignore.

This would break checksum verification on downgrades, I guess.

Reply:
What I am proposing supports downgrades. The checksum gets calculated the same way by the v1.32 reader as it was calculated by the v1.33 writer, by checksuming the `Data string` without parsing it.

Reply:
In your current code, are downgrades supported?

Reply:
> In your current code, are downgrades supported?

Most probably not as the checksum is calculated after unmarshalling JSON. Will try to fix it.

> The checksum gets calculated the same way by the v1.32 reader as it was calculated by the v1.33 writer, by checksuming the Data string without parsing it.

That wouldn't guarantee that other parts of checkpoint are not corrupted.

Reply:
What do you mean with "other parts of checkpoint are not corrupted"? Can you give an example?

I see only one purpose for the checksum: ensure that there have been no bitflips in the file while at rest. Not all filesystem have checksums for data, so I can see some (weak!) justification for having it in the application layer on top of a filesystem.